## Group SQL Practice

_Author: Matt Brems & Katie Sylvia_

### Required Preparation

In order to run SQL queries within Python/this Jupyter notebook, you will need to install the following:

```bash
conda install sqlalchemy
conda install psycopg2
```

If you get an error trying to install `psycopg2`, you can install a stand-alone package (sufficient for everything you'll need to complete this lab) by entering the following anywhere in the terminal: 

```bash 
pip install psycopg2-binary
```

(If `pip` doesn't work for you, try `pip3 install psycopg2-binary`.)

### Instructions

In this lab, you're going to have ten prompts. Each prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to get you to practice your SQL - an **important** skill in data science! While it is possible to get these answers using Pandas, it should be your goal to do everything only in SQL.

**Data**: The data used in this lab is the Iowa liquor database. It contains various tables of transactions. The schema for this database is here:

<img src="./images/schema.png" alt="schema" width="750"/>

In [2]:
# Import Pandas and Create_Engine

import pandas as pd
from sqlalchemy import create_engine

# Set up the engine to access the data.
engine = create_engine('postgresql://analytics_student:analyticsga@analyticsga-psql.generalassemb.ly:5432/iowa_liquor_sales_database')

### Prompt 1

How many rows are there in the `stores` table of the Iowa liquor database?

In [3]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT COUNT(name) FROM stores
"""

pd.read_sql_query(sql, engine)

count
0   1973

**Prompt 1 Answer**: 1973

### Prompt 2

If you sort the stores in alphabetical order by name, which store is fourth?

In [6]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT name
FROM stores
ORDER BY name
LIMIT 4
"""

pd.read_sql_query(sql, engine)

name
0  218 Fuel Express & Chubby's Liquor
1          3047 Cub Foods / Iowa City
2               3060 Cub Foods / Ames
3         3061 Cub Foods / Sioux City

**Prompt 2 Answer**: 3061 Cub Foods / Sioux City

### Prompt 3

Among those sales whose category name is `IMPORTED VODKA`, if you sort by vendor in alphabetical order, what is the bottle price and number of bottles bought in the first 3 transactions?
> You should give us three sets of numbers - one for each transaction.

In [ ]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT btl_price, bottle_qty, vendor
FROM sales
WHERE category_name = 'IMPORTED VODKA'
ORDER BY vendor
LIMIT 34

"""

pd.read_sql_query(sql, engine)

**Prompt 3 Answer**:

There are 33 transactions from vendor A Hardy/USA Ltd (second alphabetical vendor of imported vodka), leading to inconsistent results for QTY of bottles in this query. Sample quantities shown below.

1 -  $8.99	12

2 - $28.50	3

3 - $28.50	30


### Prompt 4

What is the total amount of liquor sold in February 2015?

In [24]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT SUM(total)
FROM sales
WHERE date BETWEEN '2015-02-01' AND '2015-02-28'
"""
pd.read_sql_query(sql, engine)

sum
0  21295350.55

**Prompt 4 Answer**: $ 21,295,350.55

### Prompt 5

Among all transactions where `IMPORTED VODKA` was sold, which vendor has the most transactions? Report the total number of transactions, the number of bottles sold, and the total amount of revenue from these transactions. Rename the columns to make sure there isn't any confusion as to what they mean!

In [32]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT COUNT(vendor) AS num_transactions, SUM(bottle_qty) AS bot_total, SUM(total) AS total_rev, vendor
FROM sales
WHERE category_name = 'IMPORTED VODKA'
GROUP BY vendor
ORDER BY num_transactions
DESC
LIMIT 1
"""

pd.read_sql_query(sql, engine)

num_transactions  bot_total   total_rev                            vendor
0             42338     403450  7880365.88  Pernod Ricard USA/Austin Nichols

**Prompt 5 Answer**: Pernod Ricard USA/Austin Nichols had 42,338 transactions consisting of 403,450	bottles for a total revenue of $7,880,365.88

### Prompt 6

The reason we sell things is to make money. In sales, the term **markup** means the amount of extra money charged for a product over the cost to make that product. (You can think of the markup as the profit for that product.)

In Iowa, the law states the minimum price at which liquor may be sold.

Calculate the markup (name this `Markup`) that shows the difference between the `btl_price` and the `state_btl_cost`. 

> If you haven't subtracted columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [25]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT btl_price - state_btl_cost AS markup, description, store
FROM sales
ORDER BY markup
DESC
LIMIT 10
"""

pd.read_sql_query(sql, engine)

markup                           description  store
0  $2,900.00      Cedar Ridge Barrel Proof Bourbon   2588
1  $2,900.00      Cedar Ridge Barrel Proof Bourbon   2590
2    $799.60  Absolut w/ Zing Zang Bloody Mary Mix   3447
3    $799.60  Absolut w/ Zing Zang Bloody Mary Mix   3524
4    $799.60  Absolut w/ Zing Zang Bloody Mary Mix   3944
5    $799.60  Absolut w/ Zing Zang Bloody Mary Mix   3447
6    $799.60  Absolut w/ Zing Zang Bloody Mary Mix   3354
7    $799.60  Absolut w/ Zing Zang Bloody Mary Mix   3944
8    $799.60  Absolut w/ Zing Zang Bloody Mary Mix   3354
9    $799.60  Absolut w/ Zing Zang Bloody Mary Mix   3524

### Prompt 7

Calculate the "Percentage Markup" by dividing markup by the state bottle cost. (Rather than the total profit per bottle, this will tell you how much stores are increasing their price over the state minimum.)

After calculating the "Percentage Markup," calculate the average percentage markup per **store**. Which store has the largest average markup, and what is that average markup?

In [27]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT ROUND(AVG((btl_price - state_btl_cost)/state_btl_cost)*100) as avg_markup_perc, store
FROM sales
GROUP BY store
ORDER BY avg_markup_perc
DESC
"""

pd.read_sql_query(sql, engine)

avg_markup_perc  store
0                53.0   4013
1                52.0   3833
2                52.0   4266
3                52.0   4024
4                51.0   3713
...               ...    ...
1347             50.0   2595
1348             50.0   2596
1349             50.0   2597
1350             50.0   2599
1351             50.0   2600

[1352 rows x 2 columns]

**Prompt 7 Answer**: Store 4013 had an average markup of 53% over state price

### Prompt 8

Building off of your last query, retrieve the five stores with the highest average markup percentage. Along with the store IDs and average markup percentage, please return the minimum and maximum markup percentage for those stores, as well as the names and addresses of those store.

What are the names of the five stores?

In [34]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT sales.store, 
    ROUND(AVG((btl_price - state_btl_cost)/state_btl_cost)*100) as avg_markup_perc, 
    ROUND(MIN((btl_price - state_btl_cost)/state_btl_cost)*100) as min_markup, 
    ROUND(MAX((btl_price - state_btl_cost)/state_btl_cost)*100) as max_markup,
    stores.name,
    stores.store_address
FROM sales, stores
WHERE sales.store = stores.store
GROUP BY sales.store, stores.name, stores.store_address
ORDER BY avg_markup_perc
DESC
LIMIT 5
"""

pd.read_sql_query(sql, engine)

store  avg_markup_perc  min_markup  max_markup                        name  \
0   4013             53.0        50.0       671.0      Wal-Mart 0841 / Tipton   
1   4266             52.0        50.0       671.0  Wal-Mart 1683 / Shenandoah   
2   3833             52.0        50.0       671.0    Wal-Mart 3394 / Atlantic   
3   4024             52.0        50.0       671.0  Wal-Mart 1546 / Iowa Falls   
4   2668             51.0        50.0        69.0    Hy-Vee Food Store / Leon   

                                       store_address  
0  1126 Highway 38 North\nTipton, IA 527720000\n(...  
1  705 S Fremont\nShenandoah, IA 516010000\n(40.7...  
2  1905 East 7th St\nAtlantic, IA 500220000\n(41....  
3  840 S Oak\nIowa Falls, IA 501260000\n(42.50295...  
4  1004 W 1st St\nLeon, IA 50144\n(40.73967779500...

**Prompt 8 Answer**:

Wal-Mart 0841 / Tipton<br>
Wal-Mart 1683 / Shenandoah<br>
Wal-Mart 3394 / Atlantic<br>
Wal-Mart 1546 / Iowa Falls<br>
Hy-Vee Food Store / Leon<br>

### Prompt 9

Which nine counties sell the highest number of bottles of liquor per capita (per person)? Calculate the average bottle size of the bottles of liquor sold by these counties rounded to the nearest first decimal place. Of the nine counties selling the highest number of bottles of liquor per capita, which county sells (on average) the largest bottle size?

**_BONUS FUN_**: This question can (also) be solved with a SQL subquery - that is, a query that queries a query! If you'd like to give this a try, you can read more about it [here](https://www.dofactory.com/sql/subquery). Otherwise, keep an eye out for the solution key!

In [ ]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT sales.county,  (SUM(bottle_qty)/counties.population) AS btl_per_capita, ROUND(AVG(liter_size),1) AS avg_size
FROM counties, sales
WHERE sales.county = counties.county
GROUP BY sales.county, counties.population
ORDER BY btl_per_capita
DESC
LIMIT 9
"""

pd.read_sql_query(sql, engine)

**Prompt 9 Answer**: Kossuth County (11 bottles per capita) has an average bottle size of 1.04 liters

### Prompt 10 (advanced but required!)

In the `sales` table, three columns are `btl_price`, `bottle_qty`, and `total`. The `total` for a transaction _should be_ the product of `btl_price` and `bottle_qty`. How many transactions have a value of `total` that is not equal to `btl_price` time `bottle_qty`?

> There will be a type error as well! We cannot compare type `money` to type `real`. We have to convert `total` to type `money` so that we can compare the two directly. [This link](https://docs.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-ver15#syntax) may be helpful to you!

In [68]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT count(total)
FROM sales
WHERE (total::numeric - btl_price::numeric * bottle_qty) != 0
"""

pd.read_sql_query(sql, engine)

count
0      0

**Prompt 10 Answer**: We were not able to find any sales records with a difference between total and expected total